fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -151,8 +159,13 @@ ACTOR Future<Void> queueTransactionStartRequests(
 				stats->txnSystemPriorityStartIn += req.transactionCount;
 			else if (req.priority() >= GetReadVersionRequest::PRIORITY_DEFAULT)
 				stats->txnDefaultPriorityStartIn += req.transactionCount;
-			else
+			else {
+				if (!batchRateInfo->canStart(req.transactionCount)) {
```

Comment:
This is the actual change. Rest of the commit is almost code formatting done by my editor and then the clang-format git pre-commit hook.

Reply:
This is a little tricky to reason about. The `canStart` function here takes as an argument the number of transactions already started toward the limit at any priority, but here we're only checking the size of the current batch. That means with a limit of 20, we could authorize 100 total batch transactions done 1 at a time, and based on the changes below they would be successfully started so long as there is space at the default priority. Starting these excess transactions would then subtract from the batch budget, which eventually would push the limit below 0, getting us to proper rejections. It seems like maybe that does work, though I worry that you'll end up with some weird oscillating pattern where you start a bunch of transactions all at once and then don't start any for a while.

Also, we have another weird case to consider where the limit is something like 4 and batches are coming in at size 5. Currently, this would reject every request in that case. Maybe that eventually corrects itself by the limit going up (or not, if there are enough higher priority transactions to crowd out batch ones), but I'm not certain. You could potentially remove all doubt by approving it probabilistically based on how big it is (e.g. in the scenario I described with a limit 4 and batch of 5, you could start it with 4/5 chance).